File: sdk/src/test/java/com/cxense/cxensesdk/ApiCallbackTest.java
package com.cxense.cxensesdk;

import com.cxense.cxensesdk.exceptions.CxenseException;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.api.mockito.PowerMockito.when;

/**
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2017-07-13).
 */
@RunWith(PowerMockRunner.class)
@PrepareForTest({Response.class})
public class ApiCallbackTest {
    private LoadCallback<Object> loadCallback;
    private Call<Object> call;
    private Response<Object> response;
    private ApiCallback<Object> callback;

    @Before
    public void setUp() throws Exception {
        ApiErrorParser errorParser = mock(ApiErrorParser.class);
        loadCallback = mock(LoadCallback.class);
        call = mock(Call.class);
        response = mock(Response.class);
        callback = new ApiCallback<>(loadCallback, errorParser);
        ResponseBody errorBody = mock(ResponseBody.class);
        when(response.errorBody()).thenReturn(errorBody);
        CxenseException exception = new CxenseException();
        when(errorParser.parseError(any(Response.class))).thenReturn(exception);
    }

    @Test
    public void onResponseCallbackNull() {
        callback = new ApiCallback<>(null, null);
        callback.onResponse(call, response);
        verify(response, never()).isSuccessful();
    }

    @Test
    public void onResponseSuccessful() throws Exception {
        when(response.isSuccessful()).thenReturn(true);
        when(response.body()).thenReturn(new Object());
        callback.onResponse(call, response);
        verify(loadCallback).onSuccess(any());
    }

    @Test
    public void onResponseError() throws Exception {
        when(response.errorBody()).thenThrow(new IllegalStateException());
        when(response.isSuccessful()).thenReturn(false);
        checkException(200, CxenseException.class);
    }

    @Test
    public void onFailure() throws Exception {
        CxenseException exception = new CxenseException();
        callback.onFailure(call, exception);
        verify(loadCallback).onError(exception);
    }

    private void checkException(int code, Class<? extends CxenseException> clazz) {
        when(response.isSuccessful()).thenReturn(false);
        when(response.code()).thenReturn(code);
        callback.onResponse(call, response);
        verify(loadCallback).onError(any(clazz));
    }

}

File: sdk/src/test/java/com/cxense/cxensesdk/ApiCallbackTest.kt
package com.cxense.cxensesdk

import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import retrofit2.Response
import kotlin.test.Test

class ApiCallbackTest {
    private val response: Response<Any> = mock()
    private val callback: LoadCallback<Any> = mock()
    private val apiErrorParser: ApiErrorParser = mock()
    private val apiCallback = ApiCallback(callback, apiErrorParser)

    @Test
    fun onFailure() {
        val e = Throwable()
        apiCallback.onFailure(mock(), e)
        verify(callback).onError(eq(e))
    }

    @Test
    fun onResponseSuccessful() {
        val body: Any = mock()
        whenever(response.isSuccessful).thenReturn(true)
        whenever(response.body()).thenReturn(body)
        apiCallback.onResponse(mock(), response)
        verify(response).body()
        verify(callback).onSuccess(eq(body))
    }

    @Test
    fun onResponseSuccessfulWithoutBody() {
        whenever(response.isSuccessful).thenReturn(true)
        whenever(response.body()).thenReturn(null)
        checkFailedResponse(response, BaseException())
    }

    @Test
    fun onResponseSuccessfulWithoutBodyErrorNotParsed() {
        whenever(response.isSuccessful).thenReturn(true)
        whenever(response.body()).thenReturn(null)
        checkFailedResponse(response, null)
    }

    @Test
    fun onResponseFailed() {
        whenever(response.isSuccessful).thenReturn(false)
        checkFailedResponse(response, BaseException())
    }

    @Test
    fun onResponseFailedErrorNotParsed() {
        whenever(response.isSuccessful).thenReturn(false)
        checkFailedResponse(response, null)
    }

    private fun checkFailedResponse(response: Response<Any>, exception: BaseException?) {
        whenever(apiErrorParser.parseError(response)).thenReturn(exception)
        apiCallback.onResponse(mock(), response)
        verify(response).body()
        verify(apiErrorParser).parseError(eq(response))
        verify(callback).onError(any())
    }
}
